src/backend/distributed/connection/connection_management.c

```diff
@@ -438,183 +465,373 @@ ShutdownConnection(MultiConnection *connection)
+ * connections is required) or Nothing (when the pollmode stays the same).
  */
-void
-FinishConnectionListEstablishment(List *multiConnectionList)
+static enum EventSetResult
+MultiConnectionStatePoll(MultiConnectionState *connectionState)
 {
-	ListCell *multiConnectionCell = NULL;
+	MultiConnection *connection = connectionState->connection;
+	ConnStatusType status = PQstatus(connection->pgConn);
+	PostgresPollingStatusType oldPollmode = connectionState->pollmode;
 
-	foreach(multiConnectionCell, multiConnectionList)
+	Assert(connectionState->phase == MULTI_CONNECTION_PHASE_CONNECTING);
+
+	if (status == CONNECTION_OK)
 	{
-		MultiConnection *multiConnection = (MultiConnection *) lfirst(
-			multiConnectionCell);
+		connectionState->phase = MULTI_CONNECTION_PHASE_CONNECTED;
+		return REBUILD;
+	}
+	else if (status == CONNECTION_BAD)
+	{
+		/* FIXME: retries? */
+		connectionState->phase = MULTI_CONNECTION_PHASE_ERROR;
+		return REBUILD;
+	}
+	else
+	{
+		connectionState->phase = MULTI_CONNECTION_PHASE_CONNECTING;
+	}
+
+	connectionState->pollmode = PQconnectPoll(connection->pgConn);
 
-		/* TODO: consider making connection establishment fully in parallel */
-		FinishConnectionEstablishment(multiConnection);
+	/*
+	 * FIXME: Do we want to add transparent retry support here?
+	 */
+	if (connectionState->pollmode == PGRES_POLLING_FAILED)
+	{
+		connectionState->phase = MULTI_CONNECTION_PHASE_ERROR;
+		return REBUILD;
+	}
+	else if (connectionState->pollmode == PGRES_POLLING_OK)
+	{
+		connectionState->phase = MULTI_CONNECTION_PHASE_CONNECTED;
+		return REBUILD;
+	}
+	else
+	{
+		Assert(connectionState->pollmode == PGRES_POLLING_WRITING ||
+			   connectionState->pollmode == PGRES_POLLING_READING);
 	}
+
+	return (oldPollmode != connectionState->pollmode) ? UPDATE : NONE;
 }
 
 
 /*
- * Synchronously finish connection establishment of an individual connection.
+ * WaitEventSetFromMultiConnectionStates takes a list of MultiConnectionStates and adds
+ * all sockets of the connections that are still in the connecting phase to a WaitSet,
+ * taking into account the maximum number of connections that could be added in total to
+ * a WaitSet.
  *
- * TODO: Replace with variant waiting for multiple connections.
+ * waitCount populates the number of connections added to the WaitSet in case when a
+ * non-NULL pointer is provided.
  */
-void
-FinishConnectionEstablishment(MultiConnection *connection)
+static WaitEventSet *
+WaitEventSetFromMultiConnectionStates(List *connections, int *waitCount)
 {
-	static int checkIntervalMS = 200;
+	WaitEventSet *waitEventSet = NULL;
+	ListCell *connectionCell = NULL;
+
+	/*
+	 * maxEvents is the number of events we can add to the WaitEventSet.
+	 * -3 is used to have room for the following events outside of our connection events
+	 *  - WL_POSTMASTER_DEATH
+	 *  - WL_LATCH_SET
+	 *  - pgwin32_signal_event (automatically added by CreateWaitEventSet on WIN32)
+	 */
+	const int maxEvents = FD_SETSIZE - 3;
```

Comment:
Why are we using `FD_SETSIZE` here? It is a `select()` related constant. Wait event sets doesn't seem to use selects at all.  In Windows it uses `WaitForMultipleObjects()`, which can wait for `MAXIMUM_WAIT_OBJECTS` unless the suggestions at https://docs.microsoft.com/en-us/windows/desktop/api/synchapi/nf-synchapi-waitformultipleobjects are followed. `latch.c` doesn't seem to follow those suggestions, so I'd suggest adding something like:

```c
/*
 * WaitEventSets use WaitForMultipleObjects() in windows, which can wait for up to
 * MAXMIUM_WAIT_OBJECTS events. CreateWaitEventSet() internally uses one
 * of these slots for pgwin32_signal_event, so we have (MAXMIUM_WAIT_OBJECTS - 1)
 * slots to use.
 */
#ifdef WIN32
#define MAX_EVENTSET_SIZE (MAXMIUM_WAIT_OBJECTS - 1)
#else
#define MAX_EVENTSET_SIZE 1024
#endif
```

which abstracts away the `WaitEventSet` implementation related parts, and using `MAX_EVENTSET_SIZE - 2` here.

 What do you think?

Reply:
I tried finding where I got this from, must have been an example somewhere, but couldn't find it. I went for your suggestion here.

Do you think it is safe to use the 1k hardcoded default? Or is there a header/define in postgres that would make more sense? Looking for one didn't yield much results for me.

Reply:
WaitEventSets use either `poll()` or `epoll()` in non-Windows, and these functions don't seem to have any limits unlike `select()` or `WaitForMultipleObjects()`. `1024` is rather arbitrary, we can make it a much larger number. We didn't have to do this if the implementation in `latch.c` could work around the limits of `WaitForMultipleObjects()`.